src/test/bin/winkernel/control.cpp

```diff
@@ -65,6 +65,95 @@ WDFDEVICE QuicTestCtlDevice = nullptr;
 QUIC_DEVICE_EXTENSION* QuicTestCtlExtension = nullptr;
 QUIC_TEST_CLIENT* QuicTestClient = nullptr;
 
+typedef struct QuicTestNmrClient {
+    NPI_CLIENT_CHARACTERISTICS NpiClientCharacteristics;
+    HANDLE NmrClientHandle;
+    NPI_MODULEID ModuleId;
+    KEVENT RegistrationCompleteEvent;
+    MSQUIC_NMR_DISPATCH* ProviderDispatch;
+    BOOLEAN Deleting;
+} QuicTestNmrClient;
+
+static QuicTestNmrClient NmrClient;
+
+static
+NTSTATUS
+QuicTestClientAttachProvider(
+    _In_ HANDLE NmrBindingHandle,
+    _In_ VOID *ClientContext,
+    _In_ CONST NPI_REGISTRATION_INSTANCE *ProviderRegistrationInstance
+    )
+{
+    UNREFERENCED_PARAMETER(ProviderRegistrationInstance);
+
+    NTSTATUS Status;
+    QuicTestNmrClient* Client = (QuicTestNmrClient*)ClientContext;
+    PVOID ProviderContext;
+
+    #pragma warning(suppress:6387) // _Param_(2) could be '0' - by design.
```

Comment:
Does the NMR definition need to be fixed?

Reply:
IMO, they should be.